python/cog/server/redis_queue.py

```diff
@@ -245,14 +245,22 @@ def run_prediction(
         response["output"] = None
         response["logs"] = ""
 
+        if "webhook_events" in message:
+            webhook_events = message["webhook_events"]
+            # this will raise an exception if the list isn't valid
+            WebhookEvent.validate(webhook_events)
+        else:
+            webhook_events = WebhookEvent.default_events()
+
         started_at = datetime.datetime.now()
 
         try:
             input_obj = self.InputType(**response["input"])
         except Exception as e:
             response["status"] = Status.FAILED
             response["error"] = str(e)
-            yield response
+            if WebhookEvent.DONE in webhook_events:
```

Comment:
I don't think this is a blocking observation for this change, but it feels odd to me that we have inverted control of webhook sending through the conversion of `run_prediction` into a generator method, and now are putting aspects of that back into this method.

It feels like like there might be a nicer refactoring here where all of this gets wrapped up with `WebhookSender` and a `PredictionResponse` class something like this:

```
class WebhookSender:
    def __init__(self, events_filter=None):
        self.events_filter = ...

    def send(self, event, payload):
        ...

class PredictionResponse:
    def __init__(self, data, webhook_sender):
        self.data = data
        self.webhook_sender = webhook_sender

    def start(self):
        self.data['started_at'] = ...
        self.webhook_sender.send("start", self.data)

    def append_logs(self, logs):
        self.data['logs'] += logs
        self.webhook_sender.send("logs", self.data)

    ...
```
What do you reckon?

Reply:
Yeah, that's nicer. I like your proposed refactor, though it's a bigger change than I was planning to make at the moment – I think a simpler step towards it, to keep the inversion of control we have, is to yield an `(event, response)` tuple instead of just yielding the response. That'll allow the caller to decide what to do with the information.

Refactoring into a nicer `PredictionResponse` class (and replacing most of the integration tests) can wait, I think.